torch/distributed/_composable/contract.py

```diff
@@ -0,0 +1,145 @@
+import torch
+import torch.nn as nn
+
+from collections import OrderedDict
+from typing import Any, Callable, Dict, List, Optional
+
+
+# use state_slot as key for module.__dict__ to avoid coliding with other
+# properties.
+# TODO: since all composable distributed features can share the same slot.
+class _StateKey:
+
+    # implement operator < to avoid breaking dir()
+    def __lt__(self, other: Any) -> bool:
+        return True if isinstance(other, str) else id(self) < id(other)
+
+
+class _State:
+    pass
+
+
+state_key = _StateKey()
+
+
+def contract(func):
+    r"""
+    Decorate a function as a composable distributed API, where the first
+    argument of the function must be an :class:`nn.Module` instance. The
+    decorator verifies that the wrapped function does not modify parameter,
+    buffer or sub-module fully-qualified names (FQN).
+
+    When a function ``func`` is decorated by ``@contract``, a
+    ``.state(module: nn.Module)`` method will be installed to the decorated
+    function. Then you can retrieve and modify the state on a module by calling
+    ``func.state(module)``.
+
+    Example::
+        >>> import torch.nn as nn
+        >>>
+        >>> class MyModel(nn.Module):
+        >>>     def __init__(self):
+        >>>         super().__init__()
+        >>>         self.l1 = nn.Linear(10, 10)
+        >>>         self.l2 = nn.Linear(10, 10)
+        >>>
+        >>>     def forward(self, x):
+        >>>         return self.l2(self.l1(x))
+        >>>
+        >>> @contract
+        >>> def my_feature(module: nn.Module) -> nn.Module:
+        >>>     my_feature.state(module).some_state = "any value"
+        >>>     return module
+        >>>
+        >>> model = MyModel()
+        >>> my_feature(model.l1)
+        >>> assert my_feature.state(model.l1).some_state == "any value"
+        >>> my_feature(model.l2)
+        >>> model(torch.randn(2, 10)).sum().backward()
+    """
+
+    def get_all_state(module: nn.Module) -> Dict[Callable, Dict]:
+        d = module.__dict__.setdefault(state_key, {})
+        assert isinstance(d, dict), "Distributed composable API states corrupted"
+        return d
+
+    def wrapper(module: nn.Module, *args, **kwargs) -> Optional[nn.Module]:
```

Comment:
Do we want to make `module` -> `*module`, like in the design doc?

Reply:
good point. Will update. Is it OK to update that in the follow up PR?

Reply:
Sure, not a blocker for me